// update/channel.js
const ObjectId = require("mongodb").ObjectId;
const channel = require("class/channel.js");

module.exports = async (session, query, message) => {
    if (await session.set()) {
        const { guild_id, channel_id } = query;

        if (channel.isValidId(channel_id) && await channel.isOwner(guild_id, session.user_id)) {
            const { name } = message;

            const db = await require("class/database.js");
            const result = await db.collection("guild_channels").updateOne({ _id: ObjectId(channel_id), guild_id: ObjectId(guild_id) }, { $set: { name: name } });

            if (result.modifiedCount === 1) {
                return {
                    status: true,
                    type: "TEXT_CHANNEL_UPDATE",
                    channel: {
                        id: channel_id,
                        name: name
                    }
                }
            } else {
                return {
                    status: false,
                    error: "Failed to rename emoji"
                }
            }
        } else {
            return {
                status: false,
                error: "Missing permissions"
            }
        }
    } else {
        return {
            status: false
        }
    }
}
